app/celer-g4/SensitiveDetector.cc

```diff
@@ -57,30 +63,77 @@ void SensitiveDetector::Initialize(G4HCofThisEvent* hce)
 /*!
  * Add hits to the current hit collection.
  */
-bool SensitiveDetector::ProcessHits(G4Step* step, G4TouchableHistory*)
+bool SensitiveDetector::ProcessHits(G4Step* g4step, G4TouchableHistory*)
 {
-    auto edep = step->GetTotalEnergyDeposit();
+    CELER_ASSERT(g4step);
+    auto const edep = g4step->GetTotalEnergyDeposit();  // [MeV]
 
     if (edep == 0)
     {
         return false;
     }
 
-    // Create a hit for this step
-    auto* touchable = step->GetPreStepPoint()->GetTouchable();
-    CELER_ASSERT(touchable);
+    // Insert hit
+    HitData hit;
+    {
+        auto* touchable = g4step->GetPreStepPoint()->GetTouchable();
+        CELER_ASSERT(touchable);
+        hit.id = touchable->GetVolume()->GetCopyNo();
+        hit.edep = edep;
+        hit.time = g4step->GetPreStepPoint()->GetGlobalTime();  // [ns]
+    }
 
-    unsigned int id = touchable->GetVolume()->GetCopyNo();
-    HitData data{id,
-                 edep,
-                 step->GetPreStepPoint()->GetGlobalTime(),
-                 touchable->GetTranslation()};
+    // Insert pre- and post-step data
+    StepData step;
+    {
+        auto* pre_step = g4step->GetPreStepPoint();
+        CELER_ASSERT(pre_step);
+        auto const* phys_vol = pre_step->GetPhysicalVolume();
+        CELER_ASSERT(phys_vol);
+        auto const* log_vol = phys_vol->GetLogicalVolume();
+        CELER_ASSERT(log_vol);
+
+        step.volume = log_vol->GetInstanceID();
+        step.energy_loss = edep;  // [MeV]
+        step.length = g4step->GetStepLength();  // [mm]
 
-    collection_->insert(new SensitiveHit(data));
+        // Pre- and post-step data
+        this->store_step_point(*pre_step, StepPoint::pre, step);
 
+        auto* post_step = g4step->GetPostStepPoint();  // Can be undefined
+        if (post_step)
+        {
+            this->store_step_point(*post_step, StepPoint::post, step);
+        }
+    }
+
+    collection_->insert(new SensitiveHit(hit, step));
     return true;
 }
 
+//---------------------------------------------------------------------------//
+void SensitiveDetector::store_step_point(G4StepPoint& step_point,
+                                         StepPoint point,
+                                         StepData& step)
+{
+    auto const* phys_vol = step_point.GetPhysicalVolume();
+    CELER_ASSERT(phys_vol);
+    auto const* log_vol = phys_vol->GetLogicalVolume();
+    CELER_ASSERT(log_vol);
+    auto const p = static_cast<std::size_t>(point);
+
+    step.energy[p] = step_point.GetKineticEnergy();  // [MeV]
+    step.time[p] = step_point.GetGlobalTime();  // [ns]
+    auto const& pos = step_point.GetPosition();
+    step.pos[p][0] = pos.x();  // [mm]
+    step.pos[p][1] = pos.y();  // [mm]
+    step.pos[p][2] = pos.z();  // [mm]
```

Comment:
Add a helper function for `G4ThreeVector`->`std::array`. Also since we're writing to data structures in `celeritas/io` I think we should be using celeritas units (csg) or annotating the structs.